File: hammerspoon/cr/notifier.lua
-- cr.notifier — channel-agnostic notification dispatch.
--
-- One payload shape, many delivery channels. Local channels (canvas card,
-- system notification) and remote ones (Discord/Slack/generic webhooks) all
-- register against the same interface, so the future trigger engine calls
-- notify() once and routing decides where it lands.
--
-- Presence routing: if the user has been idle past config.idleThreshold, the
-- payload is also mirrored to config.remoteChannels — deliver the reminder
-- where the user actually is, not just where the trigger fired.
--
-- payload: { title, body, icon, urgency, actions = { { label, fn }, ... }, meta }
--
-- Channels receive two arguments: `(payload, note)`. Local channels use the
-- payload, because they can call the closures in `actions`. Remote channels
-- use `note` — the cr.notification/v1 object, which is pure data and safe to
-- serialize. Keeping both means a channel never has to invent a message.

local config = require("cr.config")
local log = require("cr.log")
local ui = require("cr.notify_ui")
local notification = require("cr.notification")

local M = { channels = {} }

function M.register(name, fn)
  M.channels[name] = fn
end

-- secrets: optional cr/secrets.lua (gitignored); see secrets.example.lua
local okSecrets, secrets = pcall(require, "cr.secrets")
M.secrets = (okSecrets and type(secrets) == "table") and secrets or {}

-- Discord webhook: secrets.lua wins; otherwise reuse the DISCORD_WEBHOOK_URL
-- already configured for the Claude Code stop-hook in ~/.claude/settings.json.
local function discordWebhook()
  if M.secrets.discordWebhook then return M.secrets.discordWebhook end
  local f = io.open(os.getenv("HOME") .. "/.claude/settings.json", "r")
  if not f then return nil end
  local raw = f:read("*a")
  f:close()
  local ok, data = pcall(hs.json.decode, raw)
  if ok and type(data) == "table" and type(data.env) == "table" then
    return data.env.DISCORD_WEBHOOK_URL
  end
end

-- Telegram bot: secrets.lua only. Unlike Discord there is no existing config
-- elsewhere on this machine to fall back to, and a bot token in a repo is a
-- bot someone else owns.
local function telegram()
  local t, c = M.secrets.telegramToken, M.secrets.telegramChatId
  if t and c then return t, c end
  return nil
end

-- built-in channels ----------------------------------------------------------

M.register("card", function(p)
  ui.show({
    title = p.title, body = p.body, icon = p.icon,
    urgency = p.urgency, actions = p.actions,
    sticky = p.sticky, duration = p.duration,
  })
end)

-- optional mirror; buttons require Hammerspoon notifications set to "Alerts"
-- in System Settings, so this channel makes no interactivity promises
M.register("system", function(p)
  hs.notify.new(function()
    log.append({ event = "notify.system.clicked", title = p.title })
  end, {
    title = p.title or "",
    informativeText = p.body or "",
    withdrawAfter = 0,
  }):send()
end)

M.register("discord", function(p, note)
  local url = discordWebhook()
  if not url then
    log.append({ event = "notify.discord.skipped", reason = "no webhook configured" })
    return
  end
  local body = hs.json.encode({ content = notification.toText(note, { bold = "**" }) })
  hs.http.asyncPost(url, body, { ["Content-Type"] = "application/json" }, function(code)
    log.append({ event = "notify.discord.sent", status = code, title = p.title })
  end)
end)

M.register("slack", function(p, note) -- Slack incoming webhook
  local url = M.secrets.slackWebhook
  if not url then
    log.append({ event = "notify.slack.skipped", reason = "no webhook configured" })
    return
  end
  local body = hs.json.encode({ text = notification.toText(note, { bold = "*" }) })
  hs.http.asyncPost(url, body, { ["Content-Type"] = "application/json" }, function(code)
    log.append({ event = "notify.slack.sent", status = code, title = p.title })
  end)
end)

-- Telegram is the phone channel: a bot message lands as a push notification on
-- a device you carry, which is the whole point of routing off the Mac.
M.register("telegram", function(p, note)
  local token, chat = telegram()
  if not token then
    log.append({ event = "notify.telegram.skipped", reason = "no bot token configured" })
    return
  end
  local body = hs.json.encode({
    chat_id = chat,
    text = notification.toText(note, { bold = "*" }),
    parse_mode = "Markdown",
    -- an ambient reminder is explicitly the kind that must not buzz a pocket
    disable_notification = (note.tier and note.tier.name) == "ambient",
  })
  hs.http.asyncPost(
    string.format("https://api.telegram.org/bot%s/sendMessage", token),
    body, { ["Content-Type"] = "application/json" },
    function(code, respBody)
      log.append({ event = "notify.telegram.sent", status = code, title = p.title,
                   error = code ~= 200 and (respBody or ""):sub(1, 200) or nil })
    end)
end)

M.register("webhook", function(_, note) -- generic JSON POST fanout
  for _, url in ipairs(M.secrets.webhooks or {}) do
    hs.http.asyncPost(url, hs.json.encode(note),
      { ["Content-Type"] = "application/json" }, function(code)
        log.append({ event = "notify.webhook.sent", status = code, url = url })
      end)
  end
end)

-- dispatch --------------------------------------------------------------------

function M.notify(payload, opts)
  opts = opts or {}
  local names = {}
  for _, n in ipairs(opts.channels or config.defaultChannels) do
    names[#names + 1] = n
  end
  -- opts.noMirror: this is a re-display of something already delivered, not a
  -- new event. Without it, restoring an unanswered card after a reload would
  -- re-ping Discord — the reminder would arrive on your phone a second time
  -- because Hammerspoon restarted, which is a notification about nothing.
  if not opts.noMirror
     and config.remoteWhenAway and hs.host.idleTime() >= config.idleThreshold then
    for _, n in ipairs(config.remoteChannels or {}) do
      names[#names + 1] = n
    end
  end

  local seen, list = {}, {}
  for _, n in ipairs(names) do
    if not seen[n] then seen[n] = true; list[#list + 1] = n end
  end

  -- built once, not per channel: every channel must describe the same event,
  -- and a per-channel build would give each one its own timestamp and id
  local note = notification.build(payload, opts)

  log.append({
    event = "notify.dispatch", title = payload.title, channels = list,
    body = (payload.body or ""):sub(1, 200), icon = payload.icon,
    id = note.subject.id, referent = note.subject.bound_to,
    notification = note.id, tier = note.tier.name,
  })
  for _, n in ipairs(list) do
    local channel = M.channels[n]
    if channel then
      local ok, err = pcall(channel, payload, note)
      if not ok then
        log.append({ event = "notify.channel_failed", channel = n,
                     error = tostring(err):sub(1, 200) })
      end
    else
      log.append({ event = "notify.unknown_channel", channel = n })
    end
  end
  return list
end

-- The ways a notification can reach the user, with configuration state —
-- the web UI renders this as the per-reminder channel picker.
function M.available()
  return {
    { name = "card",    configured = true, desc = "on-screen card on this Mac (default)" },
    { name = "system",  configured = true, desc = "macOS Notification Center" },
    { name = "discord", configured = discordWebhook() ~= nil, desc = "Discord webhook" },
    { name = "slack",   configured = M.secrets.slackWebhook ~= nil, desc = "Slack webhook" },
    { name = "telegram", configured = telegram() ~= nil,
      desc = "Telegram bot — push to your phone (secrets.lua)" },
    { name = "webhook", configured = type(M.secrets.webhooks) == "table" and #M.secrets.webhooks > 0,
      desc = "generic JSON POST (secrets.lua)" },
  }
end

-- fires a sample card through the full pipeline; feedback buttons write to the log
function M.test()
  M.notify({
    title = "Test reminder",
    body = "This is what a fired reminder will look like. Buttons log feedback to the event log.",
    icon = "🧪",
    urgency = "info",
    actions = {
      { label = "Done",      fn = function() log.append({ event = "feedback", value = "done", source = "test" }) end },
      { label = "Snooze",    fn = function() log.append({ event = "feedback", value = "snooze", source = "test" }) end },
      { label = "Too early", fn = function() log.append({ event = "feedback", value = "too_early", source = "test" }) end },
    },
  })
end

return M

File: windows/crw/notifier.py
"""Channel-agnostic notification dispatch — port of cr/notifier.lua.

One payload shape, many delivery channels; register(name, fn) adds one.
Channels receive (payload, note): the payload carries callables and is what
the card uses; note is the cr.notification/v1 object — pure data, built ONCE
per dispatch so every channel describes the same event with the same id.

Presence routing: idle past the threshold mirrors delivery to the remote
channels — the reminder lands where the user actually is.

Secrets come from crw/secrets.py (gitignored; see secrets_example.py) or
environment variables — deliberately not from another tool's config file.
"""

from __future__ import annotations

import json
import os
import subprocess
import threading
import urllib.request

from . import cards, config, eventlog, notification, observer

channels: dict = {}


def register(name: str, fn) -> None:
    channels[name] = fn


try:
    from . import secrets as _secrets
except ImportError:
    class _secrets:  # type: ignore
        pass


def _secret(attr: str, env: str):
    return getattr(_secrets, attr, None) or os.environ.get(env)


def _post_json(url: str, obj: dict, tag: str) -> None:
    """Fire-and-forget JSON POST on a worker thread — a slow webhook must
    never stall the tk loop a card is drawn on."""
    def run():
        try:
            req = urllib.request.Request(
                url, json.dumps(obj).encode(),
                headers={"Content-Type": "application/json"})
            with urllib.request.urlopen(req, timeout=10) as resp:
                eventlog.append({"event": f"notify.{tag}.sent", "status": resp.status})
        except Exception as e:
            eventlog.append({"event": f"notify.{tag}.failed", "error": str(e)[:200]})
    threading.Thread(target=run, daemon=True).start()


# built-in channels ----------------------------------------------------------

register("card", lambda p, note: cards.show(p))


def _system(p, note):
    """Windows toast via PowerShell + WinRT — no dependencies, and unlike the
    macOS side there is no notification-style setting to fight. Payload goes
    over as -EncodedCommand so quoting can't break it."""
    import base64
    ps = """
[Windows.UI.Notifications.ToastNotificationManager, Windows.UI.Notifications, ContentType=WindowsRuntime] | Out-Null
$xml = [Windows.UI.Notifications.ToastNotificationManager]::GetTemplateContent([Windows.UI.Notifications.ToastTemplateType]::ToastText02)
$t = $xml.GetElementsByTagName('text')
$t.Item(0).AppendChild($xml.CreateTextNode(%s)) > $null
$t.Item(1).AppendChild($xml.CreateTextNode(%s)) > $null
$app = '{1AC14E77-02E7-4E5D-B744-2EB1AE5198B7}\\WindowsPowerShell\\v1.0\\powershell.exe'
[Windows.UI.Notifications.ToastNotificationManager]::CreateToastNotifier($app).Show([Windows.UI.Notifications.ToastNotification]::new($xml))
""" % (repr(p.get("title") or "Reminder").replace('"', "'"),
       repr((p.get("body") or "")[:200]).replace('"', "'"))
    enc = base64.b64encode(ps.encode("utf-16-le")).decode()

    def run():
        try:
            subprocess.run(["powershell", "-NoProfile", "-EncodedCommand", enc],
                           creationflags=0x08000000, timeout=15)  # CREATE_NO_WINDOW
            eventlog.append({"event": "notify.system.sent", "title": p.get("title")})
        except Exception as e:
            eventlog.append({"event": "notify.system.failed", "error": str(e)[:200]})
    threading.Thread(target=run, daemon=True).start()


register("system", _system)


def _discord(p, note):
    url = _secret("discord_webhook", "DISCORD_WEBHOOK_URL")
    if not url:
        eventlog.append({"event": "notify.discord.skipped", "reason": "no webhook configured"})
        return
    _post_json(url, {"content": notification.to_text(note, bold="**")}, "discord")


def _slack(p, note):
    url = _secret("slack_webhook", "SLACK_WEBHOOK_URL")
    if not url:
        eventlog.append({"event": "notify.slack.skipped", "reason": "no webhook configured"})
        return
    _post_json(url, {"text": notification.to_text(note, bold="*")}, "slack")


def _telegram(p, note):
    token = _secret("telegram_token", "TELEGRAM_BOT_TOKEN")
    chat = _secret("telegram_chat_id", "TELEGRAM_CHAT_ID")
    if not (token and chat):
        eventlog.append({"event": "notify.telegram.skipped", "reason": "no bot token configured"})
        return
    _post_json(f"https://api.telegram.org/bot{token}/sendMessage", {
        "chat_id": chat,
        "text": notification.to_text(note, bold="*"),
        "parse_mode": "Markdown",
        # an ambient reminder is explicitly the kind that must not buzz a pocket
        "disable_notification": (note.get("tier") or {}).get("name") == "ambient",
    }, "telegram")


def _webhook(p, note):
    for url in getattr(_secrets, "webhooks", None) or []:
        _post_json(url, note, "webhook")


register("discord", _discord)
register("slack", _slack)
register("telegram", _telegram)
register("webhook", _webhook)


# dispatch -------------------------------------------------------------------

def notify(payload: dict, opts: dict | None = None) -> list[str]:
    opts = opts or {}
    names = list(opts.get("channels") or config.DEFAULT_CHANNELS)
    # no_mirror: a re-display of something already delivered, not a new event
    # — restoring a card after a restart must not re-ping a phone
    if (not opts.get("no_mirror") and config.REMOTE_WHEN_AWAY
            and observer.idle_seconds() >= config.IDLE_THRESHOLD):
        names += config.REMOTE_CHANNELS

    seen, ordered = set(), []
    for n in names:
        if n not in seen:
            seen.add(n)
            ordered.append(n)

    note = notification.build(payload, opts)
    eventlog.append({
        "event": "notify.dispatch", "title": payload.get("title"),
        "channels": ordered, "body": (payload.get("body") or "")[:200],
        "icon": payload.get("icon"), "id": note["subject"]["id"],
        "referent": note["subject"]["bound_to"], "notification": note["id"],
        "tier": note["tier"]["name"],
    })
    for n in ordered:
        fn = channels.get(n)
        if not fn:
            eventlog.append({"event": "notify.unknown_channel", "channel": n})
            continue
        try:
            fn(payload, note)
        except Exception as e:
            eventlog.append({"event": "notify.channel_failed", "channel": n,
                             "error": str(e)[:200]})
    return ordered


def available() -> list[dict]:
    return [
        {"name": "card", "configured": True, "desc": "on-screen card on this PC (default)"},
        {"name": "system", "configured": True, "desc": "Windows toast notification"},
        {"name": "discord", "configured": bool(_secret("discord_webhook", "DISCORD_WEBHOOK_URL")),
         "desc": "Discord webhook"},
        {"name": "slack", "configured": bool(_secret("slack_webhook", "SLACK_WEBHOOK_URL")),
         "desc": "Slack webhook"},
        {"name": "telegram",
         "configured": bool(_secret("telegram_token", "TELEGRAM_BOT_TOKEN")),
         "desc": "Telegram bot — push to your phone"},
        {"name": "webhook",
         "configured": bool(getattr(_secrets, "webhooks", None)),
         "desc": "generic JSON POST (secrets.py)"},
    ]


def test() -> None:
    notify({
        "title": "Test reminder",
        "body": "This is what a fired reminder will look like. "
                "Buttons log feedback to the event log.",
        "icon": "🧪",
        "urgency": "info",
        "actions": [
            {"id": "done", "label": "Done",
             "fn": lambda: eventlog.append({"event": "feedback", "value": "done",
                                            "source": "test"})},
            {"id": "snooze", "label": "Snooze",
             "fn": lambda: eventlog.append({"event": "feedback", "value": "snooze",
                                            "source": "test"})},
        ],
    })
